### PR TITLE
Us007 advanced blog

### PR DESCRIPTION
**Pull request by DilshadGit**
Please make a code review

### CODE REVIEW

**Comment by nherriot** [comment]
[US007-advanced-blog](https://trello.com/c/b34xTsty/36-us007-advanced-blog)

File: noisyatom/blogs/forms.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+	publish = forms.DateField(widget=forms.SelectDateWidget)
+	class Meta:
+		model = Post
+		fields = [
+			'title',
+			'content',
+			'image',
+			'draft',
+			'publish',
+		]
```
Comment (nherriot) [inline noisyatom/blogs/forms.py]: Add a new line at the end of the file.

File: noisyatom/blogs/models.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	height_field = models.IntegerField(default=0)
+	width_field = models.IntegerField(default=0)
+
+	draft = models.BooleanField(default=False)
+	publish = models.DateField(auto_now=False, auto_now_add=False)
+
+	updated = models.DateTimeField(auto_now=False, auto_now_add=True)
+	created = models.DateTimeField(auto_now=True, auto_now_add=False)
+
+	objects = PostManager()
```
Comment (nherriot) [inline noisyatom/blogs/models.py]: the name 'objects' means nothing ? Can you put a doc string in the PostManager class to say what it is for? I have no idea what this 'objects' is for? The name is too generic. This does not look like clean code to me?

File: noisyatom/blogs/models.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		slug = new_slug
+	qs = Post.objects.filter(slug=slug).order_by('-id')
+	exists = qs.exists()
+	# is the title exist than create new title via slug
+	if exists:
+		new_slug = '%s-%s' %(slug, qs.first().id)
+		return create_slug(instance, new_slug=new_slug)
+	return slug
+'''
+that is mean before the model will be save to the db we will do something!
```
Comment (nherriot) [inline noisyatom/blogs/models.py]: Can you put a better doc string in this please?

File: noisyatom/blogs/templates/details.html [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+			{% if instance.image %}
+			<div class="img thumbnail">
+				<img src="{{instance.image.url}}" width="400" height="250" class="rounded img-responsive" id='id_img' />
+			</div>
+			{% endif %}
+		</div>
+
+	</div>
+</div>	
+{% endblock content %}
```
Comment (nherriot) [inline noisyatom/blogs/templates/details.html]: File needs a return character at the end.

File: noisyatom/blogs/templates/list.html [hunk 5]
```diff
@@ -0,0 +1,93 @@
+{% extends 'background-page.html' %}
+
+{% load bootstrap4 %}
+{% load urlify %}
+
+{% block content %}
+
+<div class="row">
+
```
Comment (nherriot) [inline noisyatom/blogs/templates/list.html]: Remove extra line please.

File: noisyatom/blogs/templates/list.html [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+			
+			<a href="{% url 'blogs:detail' slug=obj.slug %}" class="">
+				<button type="button" class="btn btn-primary pull-right">More ...</button>
+			</a>	
+		</div>
+
+		<div class="col-md-4">
+				<br><br>
+				{% if obj.image %}
+					<img src="{{obj.image.url}}" class="img-thumbnail img-responsive" />
```
Comment (nherriot) [inline noisyatom/blogs/templates/list.html]: Are both classes needed?
class="img-thumbnail img-responsive"

File: noisyatom/blogs/urls.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+from django.conf.urls import url
+from django.contrib import admin
+
+
+from .views import (
+		list_post, 
+		detail_post,
+		create_post,
+        update_post,
```
Comment (nherriot) [inline noisyatom/blogs/urls.py]: This is not aligned.

File: noisyatom/blogs/views.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+	form = PostForm(request.POST or None, request.FILES or None)
+	if form.is_valid():
+		instance = form.save(commit=False) 
+		instance.user = request.user
+		instance.save()
+		messages.success(request, 'The page successfully created')
+		return HttpResponseRedirect(instance.get_absolute_url())
+	else:
+		messages.success(request, 'The page not created')
```
Comment (nherriot) [inline noisyatom/blogs/views.py]: Can you change to 'The page is not created' or 'This page has not been created' or 'This page failed to create'. 
Makes more sense in English... :-)

File: noisyatom/blogs/views.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+	return render(request, template_name, context)
+
+def delete_post(request, slug=None):
+	if not request.user.is_staff or not request.user.is_superuser:
+		raise Http404
+	instance = get_object_or_404(Post, slug=slug)
+	instance.delete()
+	messages.success(request, 'The post has been deleted')
+	return redirect('blogs:list')
```
Comment (nherriot) [inline noisyatom/blogs/views.py]: Need a return character at the end of the file.

File: noisyatom/config/settings.py [hunk 10]
```diff
@@ -72,11 +72,17 @@
...
-        'account',
-        'catalog',
+
+        'account.apps.AccountConfig',
+        'blogs.apps.BlogsConfig',
+        'catalog.apps.CatalogConfig',
         'cms.apps.CmsConfig',
         'qrcode',
         'ee_test',
+
```
Comment (nherriot) [inline noisyatom/config/settings.py]: Remove space.

File: noisyatom/config/settings.py [hunk 11]
```diff
@@ -72,11 +72,17 @@
         'django.contrib.messages',
         'django.contrib.staticfiles',
         'django.contrib.sites',
-        'account',
-        'catalog',
+
```
Comment (nherriot) [inline noisyatom/config/settings.py]: Remove space.

File: noisyatom/config/settings.py [hunk 12]
```diff
@@ -137,11 +143,14 @@
...
```
Comment (nherriot) [inline noisyatom/config/settings.py]: Remove space.

File: noisyatom/config/settings.py [hunk 13]
```diff
@@ -137,11 +143,14 @@
...
+        'account.apps.AccountConfig',
+        'blogs.apps.BlogsConfig',
+        'catalog.apps.CatalogConfig',
         'cms.apps.CmsConfig',
         'qrcode',
```
Comment (nherriot) [inline noisyatom/config/settings.py]: Remove space

File: noisyatom/config/urls.py [hunk 14]
```diff
@@ -1,18 +1,7 @@
-"""noisyatom URL Configuration
```
Comment (nherriot) [inline noisyatom/config/urls.py]: Why did you remove the doc string from the urls.py?

File: noisyatom/requirements/base.txt [hunk 15]
```diff
@@ -0,0 +1,14 @@
+appdirs==1.4.0
```
Comment (nherriot) [inline noisyatom/requirements/base.txt]: Please keep to using requirements.txt as the system deploys using this. Also please add details why you want certain libraries.

File: noisyatom/static/assets/css/noisy_atom.css [hunk 16]
```diff
@@ -30,3 +30,8 @@
 .alfa-body {
   background-color: #0000EE;
 }
+
+@media (min-width: 992px){
+	.col-md-offset-10 {
+    margin-left: 80.33333333%;
+}
```
Comment (nherriot) [inline noisyatom/static/assets/css/noisy_atom.css]: Needs a return line.

File: noisyatom/blogs/models.py [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+from markdown_deux import markdown
+
+
+'''
+https://docs.djangoproject.com/en/1.11/ref/models/fields/
+'''
+
+class PostManager(models.Manager):
+
+	def active(self, *args, **kwargs):
```
Comment (nherriot) [inline noisyatom/blogs/models.py]: What are *args and **kwargs for? They are not used? What does this post manager do? There are no comments to help the developer know what this is for? There are no doc strings.

File: noisyatom/blogs/models.py [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+		return super(PostManager, self).filter(
+										draft=False).filter(
+													publish__lte=timezone.now())
+
+
+def upload_image_location(instcance, filename):
```
Comment (nherriot) [inline noisyatom/blogs/models.py]: What is the purpose of this function? It seems it only creates a string! That's not good quality. Please remove this complexity and create the string in place if it's just 1 line.

File: noisyatom/blogs/models.py (hunk 3 above)
Comment (nherriot) [reply noisyatom/blogs/models.py]: Also the doc string is not indented.

File: noisyatom/blogs/tests/test_models.py [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+class PostModelTest(TestCase):
+
+	def setUp(self):
+		self.post = Post.objects.create(title='New title test', slug='first-new-title-test-created')
+
+	def create_post(self, title='Post new blog'):
+		return Post.objects.create(title=title)
+
+	def test_create_title(self):
```
Comment (nherriot) [inline noisyatom/blogs/tests/test_models.py]: This does not create a title? It looks like it is testing the string matches a post with the defined slug title. Maybe write a comment on the method?

File: noisyatom/blogs/tests/test_models.py [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+		self.assertTrue(obj.content == '')
+	
+	def test_create_slug(self):
+		title_one = 'The new title for slug'
+		title_two = 'The new title for example with more char'
+		'''
+		The test three is only to show that if we have two title is the same or duplicate will give an error
+		because we said in the slug unique=true 
+		'''
+		title_three = 'The new title for slug'
```
Comment (nherriot) [inline noisyatom/blogs/tests/test_models.py]: Should the doc string not be at the start of the function or class?

File: noisyatom/blogs/tests/test_models.py [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+
+		obj1 = self.create_post(title=title_one)
+		obj2 = self.create_post(title=title_two)
+		obj3 = self.create_post(title=title_three)
+
+		self.assertEqual(obj1.slug, slug_one)
+		self.assertEqual(obj2.slug, slug_two)
+		self.assertNotEqual(obj3.slug, slug_three)
+
+	def test_blog_qs(self):
```
Comment (nherriot) [inline noisyatom/blogs/tests/test_models.py]: What does 'qs' stand for?

File: noisyatom/blogs/tests/test_views.py [hunk 22]
```diff
@@ -0,0 +1,40 @@
+
+from django.core.urlresolvers import reverse
+from django.test import TestCase
+from django.utils.text import slugify
```
Comment (nherriot) [inline noisyatom/blogs/tests/test_views.py]: slugify is not used. it can be deleted

File: noisyatom/blogs/tests/test_views.py [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+		self.assertEqual(response.status_code, 200)
+
+	
+	def test_detail_views(self):
+		obj = self.create_post(title='Some new title for new test')
+		response = self.client.get(obj.get_absolute_url())
+		self.assertEqual(response.status_code, 200)
+
+
+	def test_update_views(self):
```
Comment (nherriot) [inline noisyatom/blogs/tests/test_views.py]: this test looks wrong to me. it is checking for a 404. but it should work. in other words, you should be able to write and fetch a blog.

File: noisyatom/blogs/tests/test_views.py [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+		response = self.client.get(url)
+		self.assertEqual(response.status_code, 404)
+
+
+	def test_delete_views(self):
+		obj = self.create_post(title='Some new title for new test')
+		url = reverse('blogs:delete', kwargs={'slug': obj.slug})
+		response = self.client.get(url)
+		self.assertEqual(response.status_code, 404)
+
```
Comment (nherriot) [inline noisyatom/blogs/tests/test_views.py]: This looks wrong to me too. You should be able to remove blogs.

File: noisyatom/blogs/models.py (hunk 3 above)
Comment (DilshadGit) [reply noisyatom/blogs/models.py]: done

File: noisyatom/blogs/forms.py (hunk 1 above)
Comment (DilshadGit) [reply noisyatom/blogs/forms.py]: done

File: noisyatom/blogs/templates/list.html (hunk 6 above)
Comment (DilshadGit) [reply noisyatom/blogs/templates/list.html]: No change its bootstrap3 
.img-thumbnail = to give an image a rounded 1px border appearance.
.img-responsive = use for different device like mobile

File: noisyatom/blogs/views.py (hunk 8 above)
Comment (DilshadGit) [reply noisyatom/blogs/views.py]: done

File: noisyatom/blogs/views.py (hunk 9 above)
Comment (DilshadGit) [reply noisyatom/blogs/views.py]: done

File: noisyatom/config/settings.py (hunk 10 above)
Comment (DilshadGit) [reply noisyatom/config/settings.py]: we need it to put the separate between project applications , django automatic codes and third party library

File: noisyatom/config/settings.py (hunk 11 above)
Comment (DilshadGit) [reply noisyatom/config/settings.py]: we need it to put the separate between project applications , django automatic codes and third party library

File: noisyatom/config/settings.py (hunk 12 above)
Comment (DilshadGit) [reply noisyatom/config/settings.py]: we need it to put the separate between project applications , django automatic codes and third party library

File: noisyatom/config/settings.py (hunk 13 above)
Comment (DilshadGit) [reply noisyatom/config/settings.py]: we need it to put the separate between project applications , django automatic codes and third party library

File: noisyatom/config/urls.py (hunk 14 above)
Comment (DilshadGit) [reply noisyatom/config/urls.py]: We don't need this comments any more we already done what is saying in the doc string.

File: noisyatom/requirements/base.txt (hunk 15 above)
Comment (DilshadGit) [reply noisyatom/requirements/base.txt]: Changed back to as it was look like before, done

File: noisyatom/static/assets/css/noisy_atom.css (hunk 16 above)
Comment (DilshadGit) [reply noisyatom/static/assets/css/noisy_atom.css]: done

File: noisyatom/blogs/models.py (hunk 18 above)
Comment (DilshadGit) [reply noisyatom/blogs/models.py]: I have added some comments to read

File: noisyatom/blogs/models.py (hunk 17 above)
Comment (DilshadGit) [reply noisyatom/blogs/models.py]: OK I will put some comments as doc string

File: noisyatom/blogs/tests/test_models.py (hunk 20 above)
Comment (DilshadGit) [reply noisyatom/blogs/tests/test_models.py]: The reason of this doc string is about the third object created in the test, please have another look

File: noisyatom/blogs/tests/test_models.py (hunk 21 above)
Comment (DilshadGit) [reply noisyatom/blogs/tests/test_models.py]: I have changed to test_blog_queryset

File: noisyatom/blogs/tests/test_views.py (hunk 22 above)
Comment (DilshadGit) [reply noisyatom/blogs/tests/test_views.py]: Correct I have removed

File: noisyatom/blogs/templates/details.html (hunk 4 above)
Comment (DilshadGit) [reply noisyatom/blogs/templates/details.html]: done

**Comment by DilshadGit** [comment]
All fixed
